src/javascripts/ng-admin/Main/component/directive/maNgAdmin.js

```diff
@@ -0,0 +1,18 @@
+function maNgAdminDirective(NgAdminConfiguration) {
+    'use strict';
+
+    return {
+        scope: {
+            'configuration': '&'
```

Comment:
I somehow managed to pass the configuration from the new directive to controllers down the ui-view tree. 



The `<ma-ng-admin>` directive controller activates a new state called "initialize", passing it the configuration from the scope as view parameter:

```js
function maNgAdminDirective(NgAdminConfiguration, $state) {
    'use strict';

    return {
        scope: {
            'configuration': '&'
        },
        restrict: 'E',
        link: function(scope) {
            $state.go('initialize', { configuration: scope.configuration() || NgAdminConfiguration })
        },
        template: `<div ui-view></div>`
    };
}

maNgAdminDirective.$inject = ['NgAdminConfiguration', '$state'];

export default maNgAdminDirective;
```

Here is the new state definition:

```js
    $stateProvider.state('main', {
        abstract: true,
        templateProvider: ['NgAdminConfiguration', function(configuration) {
            return configuration.layout() || layoutTemplate;
        }],
        data: {
            adminConfig: {}
        }
    });
    $stateProvider.state('initialize', {
        parent: 'main',
        params: {
            configuration: null
        },
        controller: ['$state', '$stateParams', function($state, $stateParams) {
            $state.current.data.adminConfig = $stateParams.configuration;
            $state.go('dashboard');
        }]
    });
```

The only purpose of this `initialize` state is to populate the state `data` (which is declared on the `main` state, and therefore accessible to all child states`). So now the configuration is accessible to all child view:

```js
    $stateProvider.state('dashboard', {
        parent: 'main',
        url: '/dashboard?sortField&sortDir',
        params: {
            sortField: null,
            sortDir: null
        },
        controller: 'DashboardController',
        controllerAs: 'dashboardController',
        templateProvider: function() {
            return adminConfig.dashboard().template() || dashboardTemplate;
        },
        resolve: {
            dataStore: dataStoreProvider(),
            adminConfig: ['$state', function($state) {
                return $state.current.data.adminConfig;
            }],
            hasEntities: ['adminConfig', function(adminConfig) {
                return adminConfig.entities.length > 0;
            }],
            collections: ['adminConfig', function(adminConfig) {
                return adminConfig.dashboard().collections();
            }],
            // etc
```

It's convoluted, but it works. 

But: we're still injecting `NgAdminConfiguration` to a few places where I think it'll be hard to take them out:

```
> git grep NgAdminConfiguration
src/javascripts/ng-admin/Crud/column/maColumn.js:60:maColumn.$inject = ['$state', '$compile', 'NgAdminConfiguration', 'FieldViewConfiguration'];
src/javascripts/ng-admin/Crud/column/maReferenceManyLinkColumn.js:32:    maReferenceManyLinkColumn.$inject = ['$state', 'NgAdminConfiguration'];
src/javascripts/ng-admin/Crud/form/FormController.js:129:FormController.$inject = ['$scope', '$state', 'WriteQueries', 'NgAdminConfiguration', 'progression', 'notification', 'view', 'dataStore'];
src/javascripts/ng-admin/Crud/repository/ReadQueries.js:16:ReadQueriesService.$inject = ['RestWrapper', 'PromisesResolver', 'NgAdminConfiguration'];
src/javascripts/ng-admin/Crud/repository/WriteQueries.js:16:WriteQueriesService.$inject = ['RestWrapper', 'PromisesResolver', 'NgAdminConfiguration'];
```